src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageServerCustomMessageTarget.cs

```diff
@@ -1053,73 +1054,96 @@ private static TextEdit BuildRevertedEdit(TextEdit provisionalTextEdit)
 
         public override async Task<WorkspaceEdit?> RenameAsync(DelegatedRenameParams request, CancellationToken cancellationToken)
         {
-            var hostDocumentUri = request.HostDocument.Uri;
-            if (!_documentManager.TryGetDocument(hostDocumentUri, out var documentSnapshot))
+            var delegationDetails = await GetProjectedRequestDetailsAsync(request, cancellationToken).ConfigureAwait(false);
+            if (delegationDetails is null)
             {
                 return null;
             }
 
-            string languageServerName;
-            Uri projectedUri;
-            VirtualDocumentSnapshot virtualDocumentSnapshot;
-            if (request.ProjectedKind == RazorLanguageKind.Html &&
-                documentSnapshot.TryGetVirtualDocument<HtmlVirtualDocumentSnapshot>(out var htmlVirtualDocument))
-            {
-                languageServerName = RazorLSPConstants.HtmlLanguageServerName;
-                projectedUri = htmlVirtualDocument.Uri;
-                virtualDocumentSnapshot = htmlVirtualDocument;
-            }
-            else if (request.ProjectedKind == RazorLanguageKind.CSharp &&
-                documentSnapshot.TryGetVirtualDocument<CSharpVirtualDocumentSnapshot>(out var csharpVirtualDocument))
+            var renameParams = new RenameParams()
```

Comment:
Teeechnically, if we had an abstract `CreateProjectedParameters` on `DelegatedRenameParams` / `IDelegatedParams` you could bake the creation of these params into the delegated params and then the entire message target could be a single method that had a `<TParams, JToken>` call signature.

Just a spicey thought 😄

Reply:
Methods on DTOs? No thanks.